From 11bbcad50acacb4a7c36c38cc486e64332022ac1 Mon Sep 17 00:00:00 2001
From: Edward Hope-Morley <edward.hope-morley@canonical.com>
Date: Thu, 18 Jun 2015 13:38:58 +0100
Subject: [PATCH] Add support for syslog connect retries

If we have requested logging to syslog and syslog is
not yet ready we shoudl allow for retry attempts. This
patch provides a new option syslog-connect-retries to
allow for retries with a 5 second interval between
each retry.

Closes-Bug: 1459046

Change-Id: I3e5e892ce68a15f783a7d8a1a1fd56f6a53dd0ad
Co-authored-by: Liang Chen <liang.chen@canonical.com>
---
 nova/openstack/common/log.py | 49 ++++++++++++++++++++++++++++++++++----------
 1 file changed, 38 insertions(+), 11 deletions(-)

diff --git a/nova/openstack/common/log.py b/nova/openstack/common/log.py
index 5b8c685..36274c5 100644
--- a/nova/openstack/common/log.py
+++ b/nova/openstack/common/log.py
@@ -35,6 +35,7 @@ import logging.handlers
 import os
 import socket
 import sys
+import time
 import traceback
 
 from oslo.config import cfg
@@ -103,6 +104,10 @@ logging_cli_opts = [
                 help='Use syslog for logging. '
                      'Existing syslog format is DEPRECATED during I, '
                      'and will change in J to honor RFC5424.'),
+    cfg.IntOpt('syslog-connect-retries',
+               default=3,
+               help='Number of attempts with a five second interval to retry '
+                    'connecting to syslog. (if use-syslog=True)'),
     cfg.BoolOpt('use-syslog-rfc-format',
                 # TODO(bogdando) remove or use True after existing
                 #    syslog format deprecation in J
@@ -544,19 +549,41 @@ def _setup_logging_from_conf(project, version):
             logger.setLevel(level_name)
 
     if CONF.use_syslog:
-        try:
-            facility = _find_facility_from_conf()
-            # TODO(bogdando) use the format provided by RFCSysLogHandler
-            #   after existing syslog format deprecation in J
-            if CONF.use_syslog_rfc_format:
-                syslog = RFCSysLogHandler(facility=facility)
+        retries = CONF.syslog_connect_retries
+        syslog_ready = False
+        while True:
+            try:
+                facility = _find_facility_from_conf()
+                # TODO(bogdando) use the format provided by RFCSysLogHandler
+                #   after existing syslog format deprecation in J
+                if CONF.use_syslog_rfc_format:
+                    syslog = RFCSysLogHandler(facility=facility)
+                else:
+                    syslog = logging.handlers.SysLogHandler(facility=facility)
+                log_root.addHandler(syslog)
+                syslog_ready = True
+            except socket.error:
+                if CONF.syslog_connect_retries <= 0:
+                    log_root.error(_('Connection to syslog failed and no '
+                                     'retry attempts requested'))
+                    break
+
+                if retries:
+                    log_root.info(_('Connection to syslog failed - '
+                                    'retrying in 5 seconds'))
+                    retries -= 1
+                else:
+                    log_root.error(_('Connection to syslog failed and '
+                                     'max retry attempts reached'))
+                    break
+
+                time.sleep(5)
             else:
-                syslog = logging.handlers.SysLogHandler(facility=facility)
-            log_root.addHandler(syslog)
-        except socket.error:
-            log_root.error('Unable to add syslog handler. Verify that syslog'
-                           'is running.')
+                break
 
+        if not syslog_ready:
+            log_root.error(_('Unable to add syslog handler. Verify that '
+                             'syslog is running.'))
 
 _loggers = {}
 
-- 
1.9.1

